<a href="https://colab.research.google.com/github/AMaharj/AI/blob/main/Assignment_4/Maharjan_Alice_Assignment_4.ipynb\" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier

# Part 1: Load the dataset

In [2]:
# Load the dataset (load remotely, not locally)

# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)


In [3]:
iris = datasets.load_iris()
df = pd.DataFrame(iris.data, columns = iris.feature_names)
df.head(15)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                 5.1               3.5                1.4               0.2
1                 4.9               3.0                1.4               0.2
2                 4.7               3.2                1.3               0.2
3                 4.6               3.1                1.5               0.2
4                 5.0               3.6                1.4               0.2
5                 5.4               3.9                1.7               0.4
6                 4.6               3.4                1.4               0.3
7                 5.0               3.4                1.5               0.2
8                 4.4               2.9                1.4               0.2
9                 4.9               3.1                1.5               0.1
10                5.4               3.7                1.5               0.2
11                4.8               3.4                1.6               0.2
12                4.8               3.0                1.4               0.1
13                4.3               3.0                1.1               0.1
14                5.8               4.0                1.2               0.2

In [4]:
df.dtypes


sepal length (cm)    float64
sepal width (cm)     float64
petal length (cm)    float64
petal width (cm)     float64
dtype: object

In [5]:
df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)  
count        150.000000  
mean           1.199333  
std            0.762238  
min            0.100000  
25%            0.300000  
50%            1.300000  
75%            1.800000  
max            2.500000

In [6]:
df.info

<bound method DataFrame.info of      sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]>

In [7]:

iris.target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

In [8]:
iris.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

The data set consist of four freatures, sepal length, sepal widthm perap length and petal width. The label(target) is categorized in three classes, Setosa(0), versicolor(1) and virginica(2)

# Part 2: Split the dataset into train and test

In [9]:
# Take the dataset and split it into our features (X) and label (y)
x = iris.data
y = iris.target

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.10)



# Part 3: Logistic Regression

In [10]:
# i. Use sklearn to train a LogisticRegression model on the training set

# ii. For a sample datapoint, predict the probabilities for each possible class


In [11]:
Model = LogisticRegression()
Model.fit(x_train, y_train)

LogisticRegression()

In [12]:
Prediction = Model.predict_proba(x_test)

In [13]:
print({"Sample": y_test, "Predicted": Prediction})

{'Sample': array([0, 0, 2, 0, 1, 2, 1, 2, 0, 2, 1, 0, 0, 1, 1]), 'Predicted': array([[9.70511144e-01, 2.94887767e-02, 7.93450946e-08],
       [9.85088218e-01, 1.49117640e-02, 1.75292520e-08],
       [1.10986943e-04, 1.20615292e-01, 8.79273721e-01],
       [9.79125520e-01, 2.08744405e-02, 3.94878339e-08],
       [7.66851845e-04, 5.66532465e-01, 4.32700683e-01],
       [6.20844549e-06, 5.29992459e-02, 9.46994546e-01],
       [1.56255734e-02, 8.84062007e-01, 1.00312419e-01],
       [1.34144294e-06, 2.97356596e-02, 9.70262999e-01],
       [9.68763811e-01, 3.12359307e-02, 2.58375485e-07],
       [5.17034973e-07, 1.25742431e-02, 9.87425240e-01],
       [3.93980961e-03, 6.92166248e-01, 3.03893943e-01],
       [9.67152153e-01, 3.28477064e-02, 1.40233966e-07],
       [9.73218141e-01, 2.67817799e-02, 7.88529466e-08],
       [1.98271562e-02, 9.25781603e-01, 5.43912403e-02],
       [2.55907915e-03, 7.91018391e-01, 2.06422529e-01]])}


In [14]:

# iii. Report on the score for Logistic regression model, what does the score measure?

# iv. Extract the coefficents and intercepts for the boundary line(s)

In [15]:
Score = Model.score(x_test,y_test)
print("Score =", Score)

Score = 1.0


The Score measures the goodness of fit for the logistic regression models. This value shows us the relation between our features and the label of the trained model. The value ranges from 0 to 1 and higher value means better relationship between features and labels

Our value of 1.0 means that the model will always predict with 100 percent accuracy

In [16]:
Coeff = Model.coef_
Intercept = Model.intercept_
print("Coeffecient = ", Coeff)
print("Intercept = ", Intercept)



Coeffecient =  [[-0.4008652   0.94783392 -2.44891233 -1.05961672]
 [ 0.50719841 -0.30564672 -0.19260951 -0.88218872]
 [-0.10633321 -0.6421872   2.64152184  1.94180544]]
Intercept =  [  9.43146012   2.12733121 -11.55879133]


# Part 4: Support Vector Machine

In [17]:
# i. Use sklearn to train a Support Vector Classifier on the training set

# ii. For a sample datapoint, predict the probabilities for each possible class

# iii. Report on the score for the SVM, what does the score measure?

In [18]:
SupVClass = svm.SVC(probability = True)
SupVClass.fit(x_train, y_train)
Prediction = SupVClass.predict_proba(x_test)
print({"Sample": y_test, "Predicted": Prediction})

{'Sample': array([0, 0, 2, 0, 1, 2, 1, 2, 0, 2, 1, 0, 0, 1, 1]), 'Predicted': array([[0.96398064, 0.02567137, 0.01034799],
       [0.97873478, 0.01293225, 0.00833297],
       [0.0105854 , 0.03872989, 0.95068472],
       [0.9728645 , 0.0178541 , 0.00928141],
       [0.0091048 , 0.49231072, 0.49858448],
       [0.00754202, 0.00416428, 0.98829371],
       [0.00867225, 0.97143851, 0.01988923],
       [0.00964031, 0.00256618, 0.98779351],
       [0.95273518, 0.03460452, 0.0126603 ],
       [0.01147441, 0.00195985, 0.98656574],
       [0.00951379, 0.81632066, 0.17416555],
       [0.95658941, 0.03022516, 0.01318542],
       [0.96513814, 0.0244287 , 0.01043317],
       [0.00884226, 0.98176317, 0.00939457],
       [0.00699957, 0.88035999, 0.11264044]])}


In [19]:
Score = SupVClass.score(x_test,y_test)
print("Score =", Score)

Score = 1.0


The Score measures the goodness of fit for the SVC . This value shows us the relation between our features and the label of the trained model. The value ranges from 0 to 1 and higher value means better relationship between features and labels

Our value of 1.0 means that the model will always predict with 100 percent accuracy

# Part 5: Neural Network

In [20]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set

# ii. For a sample datapoint, predict the probabilities for each possible class

# iii. Report on the score for the Neural Network, what does the score measure?

# iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)

In [21]:
MLP_Class = MLPClassifier(hidden_layer_sizes = (12,6), max_iter = 1000)
MLP_Class.fit(x_train, y_train)

C:\Users\alice\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:614: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


MLPClassifier(hidden_layer_sizes=(12, 6), max_iter=1000)

In [22]:
Prediction = MLP_Class.predict_proba(x_test[:1])
print({"Sample": y_test, "Predicted": Prediction})

{'Sample': array([0, 0, 2, 0, 1, 2, 1, 2, 0, 2, 1, 0, 0, 1, 1]), 'Predicted': array([[9.87815282e-01, 1.15200632e-02, 6.64655189e-04]])}


In [23]:
Score = MLP_Class.score(x_test,y_test)
print("Score =", Score)

Score = 1.0


Our value of 1.0 means that the model will always predict with 100 percent accuracy

In [24]:
MLP_Class_500 = MLPClassifier( hidden_layer_sizes = (12,6), max_iter = 500)
MLP_Class_500.fit(x_train, y_train)
Prediction = MLP_Class_500.predict_proba(x_test[:1])
print("Predicted", Prediction)
Score = MLP_Class_500.score(x_test,y_test)
print("Score =", Score)

Predicted [[0.669704   0.22630334 0.10399266]]
Score = 0.6666666666666666


C:\Users\alice\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:614: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(


In [25]:
MLP_Class_800 = MLPClassifier(hidden_layer_sizes = (12,6), max_iter = 800)
MLP_Class_800.fit(x_train, y_train)
Prediction = MLP_Class_800.predict_proba(x_test[:1])
print("Predicted", Prediction)
Score = MLP_Class_800.score(x_test,y_test)
print("Score =", Score)

Predicted [[9.77474393e-01 2.25255885e-02 1.89392057e-08]]
Score = 0.9333333333333333


With the different iterations, i was able to get the score of 0.66 on 500 iteration, 0.933 on 800 iterations and 1.0 on 1000 iterations

In [26]:
MLP_Class_lbfgs= MLPClassifier( solver = 'lbfgs', hidden_layer_sizes = (12,6), max_iter = 1000)
MLP_Class_lbfgs.fit(x_train, y_train)
Prediction = MLP_Class_lbfgs.predict_proba(x_test[:1])
print("Predicted", Prediction)
Score = MLP_Class_lbfgs.score(x_test,y_test)
print("Score =", Score)

Predicted [[1.00000000e+000 1.95916415e-095 1.13681389e-202]]
Score = 0.9333333333333333


In [27]:
MLP_Class_lbfgs= MLPClassifier( solver = 'sgd', hidden_layer_sizes = (12,6), max_iter = 1000)
MLP_Class_lbfgs.fit(x_train, y_train)
Prediction = MLP_Class_lbfgs.predict_proba(x_test[:1])
print("Predicted", Prediction)
Score = MLP_Class_lbfgs.score(x_test,y_test)
print("Score =", Score)

Predicted [[0.95667388 0.03198342 0.0113427 ]]
Score = 1.0


C:\Users\alice\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:614: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


however on the Quasi-newton method(lbgfs) and Stochastic Gradient Descent(sgd), we still got the same result for prediction for Gradient Descent but prediction of 0.933 for Quasi-Newtons method

# Part 6: K-Nearest Neighbors

In [28]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier

# ii. For a sample datapoint, predict the probabilities for each possible class

# iii. Report on the score for kNN, what does the score measure?

In [29]:
k = KNeighborsClassifier(n_neighbors= 25 )
k.fit(x_train, y_train)
Prediction = k.predict_proba(x_test)
print("Prediction =", Prediction)

Prediction = [[1.   0.   0.  ]
 [1.   0.   0.  ]
 [0.   0.12 0.88]
 [1.   0.   0.  ]
 [0.   0.52 0.48]
 [0.   0.04 0.96]
 [0.   0.96 0.04]
 [0.   0.   1.  ]
 [1.   0.   0.  ]
 [0.   0.   1.  ]
 [0.   0.68 0.32]
 [1.   0.   0.  ]
 [1.   0.   0.  ]
 [0.   1.   0.  ]
 [0.   0.76 0.24]]


In [30]:
Score = k.score(x_test, y_test)
print ("score =", Score)

score = 1.0


Our value of 1.0 means that the model will always predict with 100 percent accuracy when the n_neighbours as value of 25

In [46]:
k_1 = KNeighborsClassifier(n_neighbors= 1 )
k_1.fit(x_train, y_train)
Prediction = k_1.predict_proba(x_test)
print("Prediction =", Prediction)
Score = k_1.score(x_test, y_test)
print ("score =", Score)

Prediction = [[1. 0. 0.]
 [1. 0. 0.]
 [0. 0. 1.]
 [1. 0. 0.]
 [0. 0. 1.]
 [0. 0. 1.]
 [0. 1. 0.]
 [0. 0. 1.]
 [1. 0. 0.]
 [0. 0. 1.]
 [0. 1. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [0. 1. 0.]
 [0. 1. 0.]]
score = 0.9333333333333333


In [45]:
k_9 = KNeighborsClassifier(n_neighbors= 9 )
k_9.fit(x_train, y_train)
Prediction = k_9.predict_proba(x_test)
print("Prediction =", Prediction)
Score = k_9.score(x_test, y_test)
print ("score =", Score)

Prediction = [[1.         0.         0.        ]
 [1.         0.         0.        ]
 [0.         0.11111111 0.88888889]
 [1.         0.         0.        ]
 [0.         0.44444444 0.55555556]
 [0.         0.         1.        ]
 [0.         1.         0.        ]
 [0.         0.         1.        ]
 [1.         0.         0.        ]
 [0.         0.         1.        ]
 [0.         0.77777778 0.22222222]
 [1.         0.         0.        ]
 [1.         0.         0.        ]
 [0.         1.         0.        ]
 [0.         1.         0.        ]]
score = 0.9333333333333333


Our value of 0.93 means that the model will always predict with high accuracy when the n_neighbours as value from 1 to 5

In [47]:
k_10 = KNeighborsClassifier(n_neighbors= 10 )
k_10.fit(x_train, y_train)
Prediction = k_10.predict_proba(x_test)
print("Prediction =", Prediction)
Score = k_10.score(x_test, y_test)
print ("score =", Score)

Prediction = [[1.  0.  0. ]
 [1.  0.  0. ]
 [0.  0.1 0.9]
 [1.  0.  0. ]
 [0.  0.5 0.5]
 [0.  0.  1. ]
 [0.  1.  0. ]
 [0.  0.  1. ]
 [1.  0.  0. ]
 [0.  0.  1. ]
 [0.  0.8 0.2]
 [1.  0.  0. ]
 [1.  0.  0. ]
 [0.  1.  0. ]
 [0.  0.9 0.1]]
score = 1.0


Our value of 0.93 means that the model will always predict with high accuracy when the n_neighbours as value from 10

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

We used four different types of classifiers in this assignment, logistic Regression, Support vector machine, neural netwodk and K nearest neighbours, Most of the models had the accuracy of 1.0 which means that the model would predict with 100 percent accuracy, which infact did surprise me, the only different predictions i received was I played around with the number of iteration and the solver for neural Network  classifier that i got the predicion of 0.66 and 0.93. For the K-Nearest neighbours as well, out of curiosity, i tried changing the values of neighbours from 1 to 10 and found out that the prediction was 0.93 from 1 to 9 and after it exceeded 10, it went up to 1.0